### PR TITLE
Bugfix/1451 blank page on unkown resource

### LINKED ISSUE #1451: Blank page in unexist statuspage slug
### ⚠️ Please verify that this bug has NOT been raised before.

- [X] I checked and didn't find similar issue

### 🛡️ Security Policy

- [X] I agree to have read this project [Security Policy](https://github.com/louislam/uptime-kuma/security/policy)

### Description

Using master branch to start up dev backend & frontend server, made a monitor an a statuspage. Then open the statuspages, changed url in address bar to an unexist statuspage slug, the page went to blank.

image

### 👟 Reproduction steps

1. Clone master branch.
2. Run `npm ci --dev`.
3. Run `npm run start-server-dev`.
4. Run `npm run dev`.
5. Open `http://localhost:3000/`.
6. Create a monitor and a status page.
7. Open the status page.
8. Change the url in address bar to an unexist statuspage slug.

### 👀 Expected behavior

Something shows `Page Not Found`-like message.

### 😓 Actual Behavior

Blank page

### 🐻 Uptime-Kuma Version

1.14.0-beta.1

### 💻 Operating System and Arch

Windows 10 Pro 21H2 x64 19044.1620

### 🌐 Browser

Microsoft Edge Dev x64 101.0.1210.2

### 🐋 Docker Version

N/A

### 🟩 NodeJS Version

v16.13.0

### 📝 Relevant log output

```shell
Both servers shows nothing.
```

### PR DESCRIPTION
**Pull request by domingospanta**
# Description

Fixes #1451

## Type of change

- Bug fix (non-breaking change which fixes an issue)

## Checklist

- [ x] My code follows the style guidelines of this project
- [ x] I ran ESLint and other linters for modified files
- [ x] I have performed a self-review of my own code and tested it
- [ x] I have commented my code, particularly in hard-to-understand areas
  (including JSDoc for methods)

## Screenshots
image

Note: I am sure if src/mixins/public.js is the best place to include the axios response interceptor, so do let me know if there another file more appropriate to hold this change.

### CODE REVIEW

File: src/mixins/public.js [hunk 1]
```diff
@@ -7,6 +7,17 @@ if (env === "development" || localStorage.dev === "dev") {
     axios.defaults.baseURL = location.protocol + "//" + location.hostname + ":3001";
 }
 
+// Add a response interceptor to catch 404 errors in dymanic routes such as /status/:slug
+axios.interceptors.response.use(function (response) {
+    return response;
+}, function (error) {
+    if(error.response.status === 404){
```
Comment (karelkryda) [inline src/mixins/public.js:14]: Please, fix lint errors

**Comment by chakflying** [comment]
I think in this case handling it within `StatusPage.vue` would be more sensible?

**Comment by karelkryda** [comment]
> I think in this case handling it within `StatusPage.vue` would be more sensible? Like adding a catch or error handler in `StatusPage.vue:541`

If there's an option to catch a 404 error globally and redirect to a 404 page, that makes more sense to me than doing it in each separate file. 

But this PR only sets it up for public pages, right? Judging by the `public.js` file.

**Comment by chakflying** [comment]
Axios is only used client-side, so it would be very weird if we later on use axios for some other client-side check and it suddenly redirected the page.

File: src/mixins/public.js (hunk 1 above)
Comment (domingospanta) [reply src/mixins/public.js]: > Please, fix lint errors

done.

**Comment by domingospanta** [comment]
> I cannot confirm or deny this.
> 
> It just made more sense to me to haggle it globally. Of course, it depends on what all this could negatively affect in the future.

Yeah, that's what I had in mind when I did this. I'd rather have this at one place(the more appropriate the better) than adding catch blocks to each axios request to do the exact same thing.

**Comment by karelkryda** [comment]
> > I cannot confirm or deny this.
> > It just made more sense to me to haggle it globally. Of course, it depends on what all this could negatively affect in the future.
> 
> Yeah, that's what I had in mind when I did this. I'd rather have this at one place(the more appropriate the better) than adding catch blocks to each axios request to do the exact same thing.

So does this work for the inner part or just for the public part?

I mean, if this captures 404 in the dashboard as well (for example, when entering a monitor ID that doesn't exist).

**Comment by domingospanta** [comment]
> > > > > I cannot confirm or deny this.
> > > > > It just made more sense to me to haggle it globally. Of course, it depends on what all this could negatively affect in the future.
> > > > 
> > > > 
> > > > Yeah, that's what I had in mind when I did this. I'd rather have this at one place(the more appropriate the better) than adding catch blocks to each axios request to do the exact same thing.
> > > 
> > > 
> > > So does this work for the inner part or just for the public part?
> > > I mean, if this captures 404 in the dashboard as well (for example, when entering a monitor ID that doesn't exist).
> > 
> > 
> > I believe it works for the public part as looking for a unexisting monitor, it shows nothing: image
> 
> I'm currently thinking about whether to display 404 even if a non-existent monitor is specified, or to treat it differently.
> 
> However, since the current solution only works for the public status page, I'm wondering if it's a good idea to treat 404 in the StatusPage.vue file or like this.

If you think it would make more sense,  I can move the code to a catch block in the StatusPage.vue file.

**Comment by louislam** [comment]
First I think I need to give some contexts here, so we could discuss in the same path:
- Dashboard = Full WebSocket
- Status Page (Public) = Axios

So there is no global solution.

And I agree with @chakflying's comment. It should not be in `public.js`, because it is a vue mixin, it actually loaded globally I believe.

In my opinion, I prefer to handle it per request rather than intercepting all 404 responses. It is because 404 could be expected sometimes for rest api. It might affect future development or lead to some unexpected behaviors.

Also I prefer to keep the current url, rather than redirecting to `/page-not-found`. I believe it will create two records in the browser history stack. When user press Back button, it go to previous page and calls the api, then it redirects to `/page-not-found` again.

**Comment by domingospanta** [comment]
> First I think I need to give some contexts here, so we could discuss in the same path:
> 
> * Dashboard = Full WebSocket
> * Status Page (Public) = Axios
> 
> So there is no global solution.
> 
> And I agree with @chakflying's comment. It should not be in `public.js`, because it is a vue mixin, it actually loaded globally I believe.
> 
> In my opinion, I prefer to handle it per request rather than intercepting all 404 responses. It is because 404 could be expected sometimes for rest api. It might affect future development or lead to some unexpected behaviors.
> 
> Also I prefer to keep the current url, rather than redirecting to `/page-not-found`. I believe it will create two records in the browser history stack. When user press Back button, it go to previous page and calls the api, then it redirects to `/page-not-found` again.

Got it. It can be on a request basis then. I will move the change to StatusPage.vue.

On the url part, the reason why I added the redirect to /page-not-found was to match the regex at router.js(142-145):
`{
        path: "/:pathMatch(.*)*",
        component: NotFound,
  }`
  This regex only works to urls next to the root "/", and this is the reason why the url /status/123 would shown a blank page instead of the component NotFound. So unless we come up with a regex to cover all dynamic urls, the solution for now would have to be to redirect to something next to the root path like /page-not-found anything unexisting uri really. (Or redering the NotFound component from StatusPage.vue which I am not sure how to do xD )

**Comment by arch1v1st** [comment]
@domingospanta - great work here; glad to see your improvements rolled into the latest version!

Unfortunately though the app still returns a HTTP status code of 200.  'Soft 404' pages are considered bad for SEO - crawlers rely on the HTTP status code for classification.  Please have a look at my writeup here, along with a link to various options.  I realize this is troublesome in a Single Page App; hope you could help devise the best way to return 'real' HTTP status codes to the clients, too!  

https://github.com/louislam/uptime-kuma/issues/759

**Comment by louislam** [comment]
> @domingospanta - great work here; glad to see your improvements rolled into the latest version!
> 
> Unfortunately though the app still returns a HTTP status code of 200. 'Soft 404' pages are considered bad for SEO - crawlers rely on the HTTP status code for classification. Please have a look at my writeup here, along with a link to various options. I realize this is troublesome in a Single Page App; hope you could help devise the best way to return 'real' HTTP status codes to the clients, too!
> 
> #759

Yes, I believe it is a modern website problem. As the router manager (vue-router) is a frontend thing, which means browser have to download the html,js,css (They are always 200) first and check the route in the frontend.

**Pull request merged**